File: legacy-submission-code-server/src/main/java/fr/gouv/stopc/submission/code/server/business/controller/KpiController.java
package fr.gouv.stopc.submission.code.server.business.controller;

import fr.gouv.stopc.submission.code.server.api.KpiApi;
import fr.gouv.stopc.submission.code.server.business.model.Kpi;
import fr.gouv.stopc.submission.code.server.business.service.KpiService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.LocalDate;
import java.util.List;

/**
 * Kpi generation REST controller definition
 * 
 * @author plant-stopcovid
 */
@RestController
@RequestMapping(value = "/internal/api/v1")
public class KpiController implements KpiApi {

    /**
     * The Kpi generation service
     */
    private KpiService kpiService;

    /**
     * Spring Injection constructor
     * 
     * @param kpiService the <code>IKpiService</code> bean instance to inject
     */
    public KpiController(KpiService kpiService) {
        this.kpiService = kpiService;
    }

    /**
     * Computes the Kpi on a period
     *
     * @param fromDate beginning date of the period
     * @param toDate   ending date of the period
     * @return the list of Kpi computed on this period (one per day)
     */
    @Override
    public ResponseEntity<List<Kpi>> kpi(LocalDate fromDate, LocalDate toDate) throws Exception {
        return ResponseEntity.ok(this.kpiService.generateKPI(fromDate, toDate));
    }
}

File: src/main/kotlin/fr/gouv/stopc/submissioncode/controller/KpiController.kt
package fr.gouv.stopc.submissioncode.controller

import fr.gouv.stopc.submissioncode.api.KpiApi
import fr.gouv.stopc.submissioncode.api.model.Kpi
import fr.gouv.stopc.submissioncode.service.KpiService
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.format.annotation.DateTimeFormat.ISO.DATE
import org.springframework.http.ResponseEntity
import org.springframework.validation.BindException
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.time.Duration
import java.time.LocalDate

@RestController
@RequestMapping("/internal/api/v1")
class KpiController(val kpiService: KpiService) : KpiApi {

    override fun kpi(@DateTimeFormat(iso = DATE) fromDate: LocalDate, @DateTimeFormat(iso = DATE) toDate: LocalDate): ResponseEntity<List<Kpi>> {
        val validation = BindException(object {}, "params")
        val period = Duration.between(fromDate.atStartOfDay(), toDate.atStartOfDay())
        if (period.toDays() < 0) {
            validation.reject("INCONSISTENT_BOUNDS", "The 'fromDate' should be before 'toDate'")
        }
        if (period.toDays() > 100) {
            validation.reject(
                "PERIOD_TOO_LARGE",
                "The period between 'fromDate' and 'toDate' must be less than 100 days"
            )
        }
        if (validation.hasErrors()) {
            throw validation
        }
        val kpis = kpiService.computeKpi(fromDate, toDate)
        return ResponseEntity.ok(kpis)
    }
}
